### ISSUE TITLE
Very slow on get cached items

### ISSUE THREAD

**Issue by aborja79** [silentbobbert/EFCache.Redis#46]
Hello, I'm using your library to caching not frecuently changed data, for example application user list, but the performance is poor.
I am trying to cache the list of application users but it takes 1 second to return the list to me.
If I execute in redis server:
get 0opCYvv1sllRpag1z0iTAisa8lQ=
redis takes 0.96s
If I execute:
debug object 0opCYvv1sllRpag1z0iTAisa8lQ=
The result is
Value at:0x7f061863c8a0 refcount:1 encoding:raw serializedlength:230241 lru:11696353 lru_seconds_idle:19

Is too large the list?
Any idea?
Thank you

**Comment by silentbobbert** [comment]
Hi @aborja79 - are you saying if you query Redis directly, it takes about 0.96 seconds? If you use my package it takes 1 second? 0.04 seconds longer?

I may have misunderstood your comments.

I would point out that my package serialises and deserialises values from strings to objects and vice versa. Depending on how many records you have and how large they are, this could explain "slowness"

Have you tried different redis servers? Could it be an infrastructure issue?

**Comment by aborja79** [comment]
If I use your package it takes about 1s to get cached items.
The query returns about 10.000 records.
It could be an infraestructure issue but I tested it in 2 different
virtualized servers.

**Comment by silentbobbert** [comment]
Maybe you could test (and time) how long it takes to deserialise 10k of those records after using the BinaryFormatter like we do in StackExchangeRedisExtensions

```
        static byte[] Serialize<T>(T o) where T : class
        {
            if (o == null)
            {
                return null;
            }

            var binaryFormatter = new BinaryFormatter();

            using (var memoryStream = new MemoryStream())
            {
                binaryFormatter.Serialize(memoryStream, o);
                var objectDataAsStream = memoryStream.ToArray();
                return objectDataAsStream;
            }
        }

        static T Deserialize<T>(byte[] stream)
        {
            if (stream == null || !stream.Any())
            {
                return default(T);
            }

            var binaryFormatter = new BinaryFormatter();
            using (var memoryStream = new MemoryStream(stream))
            {
                memoryStream.Seek(0, SeekOrigin.Begin);
                var result = (T)binaryFormatter.Deserialize(memoryStream);
                return result;
            }
        }
```
If you can see how long it takes doing this in-memory, then you might see that a large portion of time is spent in deserialisation.
If this is quick, then it points more to an infrastructure issue.
If this time is close to the 1s you saw, then I am not sure how we optimise the deserialisation further...

**Comment by aborja79** [comment]
Serialize and deserialize in MemoryStream is very fast, I tested it.
The problem is when I use redis server as cache.

Could you help me to improve or test redis server performance, please?
Currently the server is running in a virtualized debian in VMWare
infrastructure.

Redis is configured to use only memory with persistence off.

I haven't any experience with redis server.

Thank you very much.

**Comment by silentbobbert** [comment]
Two things you could try:
1. Use the redis binary server bundled in the tests project. 
2. Use a docker container to provide a redis server and port forward from localhost:redisport to dockercontainer:redisport

Both should offer reasonable performance for local testing. This will remove network IO from the equation. 

I think there are some pointers to using a docker container in the tests (as a reminder for me on how to run the tests against docker redis).

The version bundled with the tests is quite old. I think I built it from source 4-5 years ago! 

So in a nutshell, I am suggesting that you try redis running on your local machine to remove network IO from the problem.

**Comment by aborja79** [comment]
I tried to connect with local redis server and the performance is better.
I tested with 1000 serialized users.
Users table has about 80 fields.
The library takes 246ms retrieving cached list in local and 912ms in remote
server with redis windows binary too.

I think is too slow in both tests.

In production redis is running in debian linux and is slower. 1248ms.

Is that performance normal?

Thank you very much.

**Comment by silentbobbert** [comment]
Hi @aborja79 - its difficult or impossible to say what "normal" is. If it's consistent, then that is "normal" for your use case and infrastructure.

You could try comparing against another form of distributed cache, such as memcached. You could create a small console application to test the same 1000 users x 80 fields, and compare.

Did you mean that the binary (very old) supplied in the test projects, running on a remote windows server, manages to supply the data (and be deserialized on your machine) in 912ms, but the "prod" redis running on a remote linux server took 1248ms for the same test?

I would have expected a 4-5 year old Windows binary to be less efficient than a current version running on Linux.

Could you try Redis-as-a-service running in Azure/AWS? Could you try a different distributed cache (like memcached) running in a public cloud?

This library extends EFCache from being an in-memory/in-process cache to using an out-of-process distributed cache. You could consider creating a new extension for EFCache based on memcached if you find the performance of memcached superior to that of redis.

For reference, when I was just reading up on the memcached wiki, it says it should be able to return an object from cache, using the lookup key, on "slow" machines in around 1ms. It doesnt say how large that one item can be. Memcached works in a similar way to redis - objects need to be serialised before being stored.

How long does it take you to retrieve the data for 1000 users x 80 fields directly from SQL server?

**Comment by aborja79** [comment]
Yes, the old binary in Windows is faster than latest version in linux. All
running in VMWare infrastructure.

I couldn't try it in memcached because We are using redis to manage shared
sessions and signalr and I don't want use additional libraries or servers.
We haven't azure or AWS account, but latency must be slower because now all
servers is inside VMWare infrastructure.
.
If I take data directly from SQL Server it takes 12s.

Thank you.

**Comment by silentbobbert** [comment]
Hi @aborja79 
I wasnt suggesting you change your whole solution over to memcached. I was wondering if you could create a small PoC with just enough code to test your 1000 user x 80 field scenario. It would be neat if you could substitute memcached to redis with a simple change to directly compare performance.

I was also suggesting a public cloud version of memcached/redis because its quick, cheap and easy to set up. Also their offerings would probably be optimised builds. Plus, the major clouds, Azure/AWS, have very good internet egress speeds. To get the best performance, the PoC should run on the same cloud provider's infrastructure too, to eliminate as much network IO as possible.
It doesnt make sense to me that a 4-5 years old Windows binary I compiled without much consideration would be faster than an up to date version running in Linux either. Still that is 912 v 1248ms - which are within the same order of magnitude.

Do you have any idea how much data (KBs/MBs) that 1000 x 80 is? If there are many string fields of say nvarchar(256) or nvarchar(max), its possible that you are serialising and transmitting a reasonable amount of data, even if it is cached.

I would say you are still seeing a 10x improvement using Redis which although could/should be better, is still a worthy improvement.
